add `docker stack ls`

AkihiroSuda opened the pull request:
**\- What I did**
add `docker stack ls`

**\- How I did it**
Please refer to the code.
I think ideally we should have `engine-api/types/swarm.Stack` and `engine-api/client.StackList`, but I didn't do so in this PR.

**\- How to verify it**

```
$ docker stack ls
NAME
stack01
```

Note that ID is not shown because it is not defined in current engine API.

Signed-off-by: Akihiro Suda suda.akihiro@lab.ntt.co.jp

fixes https://github.com/docker/docker/issues/24255

AkihiroSuda committed:
add `docker stack ls`

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

dnephin commented on api/client/stack/common.go:
```diff
@@ -48,3 +50,43 @@ func getNetworks(
+		if !ok {
+			stack := &stack{
+				Name: name,
+			}
```
I wonder if it would be valuable to include a count of services in the stack as another field.

dnephin commented on api/client/stack/common.go:
```diff
@@ -48,3 +50,43 @@ func getNetworks(
+func getStacks(
+	ctx context.Context,
+	apiclient client.APIClient,
+) ([]*stack, error) {
```
I think this would be better in `api/client/stack/list.go`, it's specify to listing stacks.

dnephin commented:
Note that these stacks commands are experimental right now, so they may change. 

As you mentioned it would be easier to support listing stacks if they were part of the API.

AkihiroSuda commented:
Done

thaJeztah commented:
we need to decide which experimental options should be included in 1.12, and which ones should go in "after" 1.12; we're already releasing 1.12-rc3 and although there's some UX changes in the RC's for this release, we should be careful to not add all of them

icecrime commented:
Thanks @AkihiroSuda! I'm going to remove this one from the milestone to avoid overloading an already quite significant release ;-) But definitely something to look into after the release!

thaJeztah commented:
@icecrime wondering; if there's no way to get a list of stacks, would that make using (e.g.) `docker stack rm` really complicated, since you don't have a way to get the names of the stacks that exist?

I think we should reconsider adding this to 1.12 (experimental)

dnephin commented:
I think there are two common scenarios for `rm`.

1. You're working on something, and when you're done, you want to clean it up instead of leave it running. You definitely know the name of the stack because you just started it.
2. You're trying to clean up unnecessary services/stacks from a swarm. You would probably want to look at `docker service ls` first anyway, to get a full picture of all the services, and from there you'd be able to see the stack names to clean them up the stacks (instead of individual services).

I think having a `stack ls` is going to be really useful, but I also think it's something that we don't absolutely need for 1.12, and it has the potential to change quite a bit in 1.13 as we finalize the commands.

AkihiroSuda commented:
ping :smiley:

dnephin commented:
design LGTM

cpuguy83 commented:
LGTM.
Could probably be freshened up with a rebase.

AkihiroSuda commented:
rebased to upstream

AkihiroSuda commented on cli/command/stack/services.go:
```diff
@@ -16,10 +16,6 @@ import (
 )
 
-const (
-	listItemFmt = "%s\t%s\t%s\t%s\t%s\n"
```
this constant had not been used, so it doesn't conflict with new one

vdemeester commented:
Hum also it's experimental, I would really like a little bit more tests (unit, integration) on that feature 👼

AkihiroSuda commented:
@vdemeester OK, updated PR and added `TestStackWithDAB`.
Maybe other tests will be added after https://github.com/docker/docker/issues/26327

CC @dnephin @thaJeztah @icecrime @cpuguy83

cpuguy83 commented:
LGTM if janky is happy.

dnephin commented:
LGTM

vdemeester commented:
Cool, thanks @AkihiroSuda 
/cc @thaJeztah for doc review 👼

thaJeztah commented:
ping @AkihiroSuda can you update the docs? then we should be ready to merge

AkihiroSuda commented:
done, sorry for didn't added that 💦

thaJeztah commented:
Just the reference docs for the command should be ok I think

AkihiroSuda commented:
OK, (almost) reverted experimental/docker-stacks-and-bundles.md

thaJeztah reviewed (approved):
LGTM
